src/nasui/src/App.tsx

```diff
@@ -0,0 +1,358 @@
+
+const styles = createStyles({
+  bottomAppBar: {
+    top: 'auto',
+    bottom: 0,
+    zIndex: 'auto',
+  },
+  title: {
+    flexGrow: 1,
+    textAlign: 'left'
+  },
+  panel: {
+    position: 'absolute',
+    top: 76,
+    right: 16,
+    width: 400,
+  },
+  listItem: {
+    paddingLeft: 0,
+    paddingRight: 0,
+    paddingTop: 2,
+    paddingBottom: 2,
+    fontSize: '0.8em',
+    wordBreak: 'break-all',
+  },
+  listSubtitle: {
+    fontWeight: 600,
+    paddingLeft: 0,
+    paddingRight: 0,
+    fontSize: '0.9em',
+  },
+  listTitle: {
+    lineHeight: 1.1,
+    wordBreak: 'break-all'
+  },
+  backdrop: {
+    color: '#fff',
+    zIndex: 100,
+  },
+  snackbar: {
+    bottom: 76
+  }
+});
+
+const ExpansionPanel = withStyles({
+  root: {
+    '&$expanded': {
+      margin: 'auto',
+    },
+  },
+  expanded: {},
+})(MuiExpansionPanel);
+
+const ExpansionPanelSummary = withStyles({
+  root: {},
+  content: {
+    '&$expanded': {
+      margin: '12px 0',
+    },
+  },
+  expanded: {},
+})(MuiExpansionPanelSummary);
+
+const ExpansionPanelDetails = withStyles(theme => ({
+  root: {
+    paddingTop: 0,
+    paddingBottom: theme.spacing(1),
+  },
+}))(MuiExpansionPanelDetails);
+
+type AppState = {
+  graph: Graph | undefined,
+  graphData: any,
+  logData: any[],
+  sliderValue: number,
+  maxSliderValue: number,
+  sliderStep: number,
+  settingsOpen: boolean,
+  hideSidechainNodes: boolean,
+  hidePrimitiveNodes: boolean,
+  snackbarOpen: boolean,
+  selectedNode: string,
+  loading: boolean,
+  layout: boolean,
+}
+
+type AppProps = {
+  classes: any
+}
+
+class App extends React.Component<AppProps, AppState>  {
+  constructor(props: any) {
+    super(props);
+    this.state = {
+      graph: undefined,
+      graphData: undefined,
+      logData: [],
+      sliderValue: 0,
+      maxSliderValue: 0,
+      sliderStep: 1,
+      settingsOpen: false,
+      hideSidechainNodes: true,
+      hidePrimitiveNodes: true,
+      selectedNode: '',
+      loading: false,
+      snackbarOpen: false,
+      layout: false,
+    };
+    this.refresh = this.refresh.bind(this);
+  }
+
+  componentDidMount() {
+    this.refresh();
+  }
+
+  refresh() {
+    this.setState({ loading: true });
+    fetch('/refresh')
+      .then((response) => { return response.json() })
+      .then((data) => {
+        const graph = new Graph(data.graph, this.state.hideSidechainNodes);
+        this.setState({
+          graphData: data.graph,
+          graph: graph,
+          logData: data.log,
+          maxSliderValue: data.log.length - 1,
+          sliderStep: Math.max(1, Math.floor(data.log.length / 20)),
+          sliderValue: Math.min(data.log.length, this.state.sliderValue),
+          loading: false,
+          snackbarOpen: graph.nodes.length > 100
+        });
+      });
+  }
+
+  private renderExpansionPanel() {
+    const { classes } = this.props;
+    const { selectedNode, graph } = this.state;
+    if (graph === undefined)
+      return null;
```

Comment:
what is the difference between `null` and `undefined`?

Reply:
React suggests `null` when returning a None element. Fixed. Thanks.